docs/rfcs/021-metering.md

```diff
@@ -0,0 +1,175 @@
+
+With the push model, it is easier to publish data only about actively changing metrics -- pageserver knows when it performs s3 offloads, garbage collection and starts/stops consuming data from the safekeeper; proxy knows about connected clients; console / autoscaler-agent knows about active cpu time.
+
+Hence, let's go with a push-based model.
+
+## Common bus vs. proxying through the console
+
+We can implement such push systems in a few ways:
+
+a. Each component pushes its metrics to the "common bus", namely segment, Kafka, or something similar. That approach scales well, but it would be harder to test it locally, will introduce new dependencies, we will have to distribute secrets for that connection to all of the components, etc. We would also have to loop back some of the events and their aggregates to the console, as we want to show some that metrics to the user in real-time.
+
+b. Each component can call HTTP `POST` with its events to the console, and the console can forward it to the segment for later integration with metronome / orb / onebill / etc. With that approach, only the console has to speak with segment. Also since that data passes through the console, the console can save the latest metrics values, so there is no need for constant feedback of that events back from the segment.
+
+# Implementation
+
+Each (proxy|pageserver|autoscaler-agent) sends consumption events to the single endpoint in the console:
+
+```json
+POST /usage_events HTTP/1.1
+Content-Type: application/json
+
+[{
+    "metric": "s3_tenant_size_bytes",
+    "type": "absolute",
+    "tenant_id": "5d07d9ce9237c4cd845ea7918c0afa7d",
+    "time": ...,
+    "value": 12345454,
+},
+...
+]
+```
+
+![data flow](./images/metering.jpg)
+
+Events could be either:
+* `incremental` -- change in consumption since the previous event or service restart. That is `effective_cpu_seconds`, `traffic_in_bytes`, and `traffic_out_bytes`.
+* `absolute` -- that is the current value of a metric. All of the size-related metrics are absolute.
+
+Each service can post events at its own pace and bundle together data from different tenants/endpoints.
+
+The console algorithm upon receive of events could be the following:
+
+1. Create and send a segment event with the same content (possibly enriching it with tenant/timeline data for endpoint-based events).
+2. Update the latest state of per-tenant and per-endpoint metrics in the database.
+3. Check whether any of that metrics is above the allowed threshold and stop the project if necessary.
```

Comment:
For some reason I can't comment to the next two comments, so replying to them here:

> A pipeline with vector accepting events via HTTP (any protocol it supports really) + Kafka as a bus + Kafka-Snowflake connector to store raw events in Snowflake + Kafka to Postgres connector to store events in our own single RDS could be a reliable solution.

Yes, that should work, but IMO we can start with the current approach as it quite dumb and experiment with different event delivery schemes after.

Why do we need vetror in that scheme instead of just pushing events to kafka? To avoid kafka-client dependency?

I started thinking of Kafka mostly due to the Kafka streams / ksql -- we could have cheap running aggregates with them. Otherwise it look that it would be pretty similar to just inserting batch of events in snowflake+postgres right in the `usage_events` handler. But with postgres/snowflake such queries will still be expensive (both in wall time and cpu consumption).

Another alternative to Kafka streams is Materialize, I'd be interested to try it out experimentally.

Reply:
> Why do we need vetror in that scheme instead of just pushing events to kafka? To avoid kafka-client dependency?

That's what I meant, yes, to not introduce new dependencies on side of the event sources. Vector has a lot of inputs, HTTP is one of them, as long as we already introduce support for sending events via HTTP. Also, Vector can 
1. enrich events on its side, adding `region` field to the payload, for example.
2. take care of retries in cases of network problems. It has both a simple in-memory buffer and an on-disk buffer for persistence.